### ISSUE TITLE
Can't add or remove nodes intuitively using new ASLayoutSpec

### ISSUE THREAD

**Issue by aaronschubert0** [facebook/AsyncDisplayKit#792]
Hi, I've been playing round with the new layout using the Kittens example project. The new layout api is great but coming from ComponentKit, something that has stump me immediately is the ability to easily add/remove nodes. I had assumed that I could just alter the array that is passed to `ASStackLayoutSpec` like so (KittenNode.mm):
```
  [stackSpec setChildren:!_swappedTextAndImage ? @[_imageNode, _textNode] : @[_textNode, _imageNode, _secondImageNode]];
```
This works in ComponentKit, but in ASDK the node is either not removed or added but the layout isn't recalculated properly. Are there other methods to achieve this or is this supposed to work?

**Comment by nguyenhuy** [comment]
You probably needs to call `-setNeedsLayout` on the node to trigger a new layout pass. If your root node is a cell node and the new layout has a different size, you can simply call [`-relayoutRowAtIndexPath:withRowAnimation`](https://github.com/facebook/AsyncDisplayKit/blob/2c8626e1f6e99468d5fe2baa6489d4fba0ec0c0c/AsyncDisplayKit/ASTableView.mm#L480) on the table view (or `-relayoutItemAtIndexPath:` on the collection view).

**Comment by aaronschubert0** [comment]
Thanks @nguyenhuy i'm using the `-toggleNodesSwap` method in `KittenNode.mm` that calls `-setNeedsLayout` like so:
```
- (void)toggleNodesSwap
{
  _swappedTextAndImage = !_swappedTextAndImage;
  [UIView animateWithDuration:0.5 delay:0 usingSpringWithDamping:0.85 initialSpringVelocity:0 options:UIViewAnimationOptionCurveEaseInOut animations:^{
    [self setNeedsLayout];
    [self.view layoutIfNeeded];
  } completion:^(BOOL finished) {}];
}
```
Would this not be enough to trigger the layout pass?

**Comment by nguyenhuy** [comment]
It is enough. But if the new layout has a different size, the result won't be picked up and rendered until the containing table view/collection view updates. In that case you need to notify the table view, by making an empty batch update, or calling -relayoutItemAtIndexPath.

Also, how deep is your node tree? Are you adding/removing direct subnodes of the cell node? Have you put a breakpoint in -layoutSpecThatFits: and are sure that the method is called?

If above suggestions don't work for you, I guess you need to provide more information, like what behaviour are you observing (a video would be best) and maybe a related piece of our code.

**Comment by aaronschubert0** [comment]
Ah okay, that would make sense. If I call -relayoutItemAtIndexPath or rotate the screen it redraws correctly. I've put a breakpoint in -layoutSpecThatFits: and it's definitely being called, and yeah I'm adding direct subnodes of the cell node.

I think it comes down to convenience more than anything then. It would be by far nicer to be able to notify the table/collection view by calling a method on the cell node, rather than having to implement a delegate or notification back into the table/collection view and call a method there. Have you ever encountered situations where you have had to add or remove subnodes and if so how do you accomplish it?

I've attached videos for adding and removing subnodes just to document the behaviour.
![adding_sub_node](https://cloud.githubusercontent.com/assets/9281756/10817503/26d6387e-7e32-11e5-88ad-c5327769d51d.gif)
![removing_nodes](https://cloud.githubusercontent.com/assets/9281756/10817512/2e1d9294-7e32-11e5-85a0-9806425e2e87.gif)

**Comment by nguyenhuy** [comment]
Thanks for the videos. It's clear that you need to (somehow) notify the table view in this case. If I was in this situation, I would use a delegate (if the change is not originated from the table view, like [this case](https://github.com/facebook/AsyncDisplayKit/blob/master/examples/Kittens/Sample/ViewController.m#L123)). 

That's said, I agree that it is not convenient to expect developers to manually notify back to the table view. And in fact we have got a few questions related to this requirement. I think we can do better than that. What do you think if we define a new ASCellNodeDelegate that is conformed by ASTableView and ASCollectionView, then `-setNeedsLayout` of ASCellNode will automatically call the delegate (only if there is a size change)?

**Comment by aaronschubert0** [comment]
That would be perfect! That way we can call this method in the same animation block (or pass the same animation options) as `-setNeedsLayout` and have everything animate at the same time. I strongly favour this approach, and i think the community will do, since nodes can be more self contained in this way without being linked by ourselves to a certain ASTableView/ASCollectionView.

**Comment by nguyenhuy** [comment]
That sounds good. I think we can even remove `ASTableView -relayoutItemAtIndexPath:` and `ASCollectionView -relayoutRowAtIndexPath:`, since calling `ASCellNode -setNeedsLayout` will be enough in all use cases. Making a diff now, shouldn't take too long :)

**Comment by aaronschubert0** [comment]
Awesome, excited to see it!

**Comment by appleguy** [comment]
@aaronschubert0 thanks a lot for filing an issue and talking with us about this — I appreciate you taking the time to give feedback to the community!

AsyncDisplayKit is in the midst of its evolution from a very raw, high-performance tool that needs to be wielded carefully -- to a highly automatic, convenient set of behaviors that make normal app development easier.  This is a key example where we've not yet fully bridged the gap and made things convenient rather than just powerful.  Another example is the stretching text you see in the videos; this is the current default behavior (more advanced usage can customize this to suit the requirements of the animation).  A better default is probably possible.

Please let us know if you have any other experiences or feedback to share!

**Comment by aaronschubert0** [comment]
Hey @appleguy thanks a lot for taking the time to respond to me in full, really appreciate that. The AsyncDisplayKit community is really awesome. It's awesome to hear of that transition taking place right now, an exciting time to take part. It's interesting that you mentioned the stretching text, because I was wondering about this actually. A better default may be a simple fade but to be honest it's more likely that the documentation still needs to be in place so that we know what's possible, which you have mentioned elsewhere is a big addition/priority of ASDK 2.0. I believe ASDK is going to be really significant going forward so I will not hesitate to give as much feedback as possible, and generally be more involved. Since this 'issue' has been addressed in #793 I'm going to go ahead and close this now, it's amazing how quick the turnaround was!
